### DISCUSSION TITLE
API server config hot reload

### DISCUSSION THREAD

**Discussion by manasmulay** [Altinity/clickhouse-backup#968] [Q&A] [answered]
I had a use case where different tables on the same node need to have different config.yaml files. Is there a way to have the API server of the backup tool running, and provide the config per backup request?
If yes, a follow up question would be whether the changed config would reflect across all "in progress" operations (using API_ALLOW_PARALLEL = true)?
If no, is there any such feature planned for future releases ?
Thanks!

**Comment by Slach** [comment, marked as answer]
Try
```sql
INSERT INTO system.backup_actions(command) VALUES('create_remote --config=/path/to/config.yml --table=db.table backup_name');
```

you can't change `in progress` command, this is inconsistent behavior

we will not release same or similar feature

which is your use case actually? maybe we could find proper solution?

**Reply by manasmulay** [reply]
Thanks for the reply! This is really useful.
 
I will elaborate my use case further,
Say I have two ClickHouse tables T1 and T2 on the same machine. I want to backup T1 to (Amazon) S3 and T2 via rsync. Since the backup stores are different for these tables, I would require different configs for every backup request. I want to have the backup tool running in API mode. Now if I issue a backup action for T1 (S3), would I be able to issue another backup request for T2 (which backs-up via rsync) by updating the db entry per the above query? Backup for T2 could be issued while T1 is being backed up.

My major concern is - would the config update with the above insert query in any way affect any ongoing backup for T1 ?

Thanks in advance!

**Reply by Slach** [reply]
try latest version
INSERT INTO system.backup_actions('create_remote --config=config-for-t1.yaml --tables=db.t1 t1_backup_name');
INSERT INTO system.backup_actions('create_remote --config=config-for-t2.yaml --tables=db.t2 t2_backup_name');

**Reply by manasmulay** [reply]
This worked, thanks! I was able to use different configs per backup.

However, I tried the following
- Start the API server with a correct config with allow_parallel = true
- Proceed to insert a new backup_action with a faulty config (say incorrect auth for ch server)
- Kill the command with the faulty config via API (specify command in the kill API, using html encoding)
- Create a new backup using /backup/create api
The last backup action succeeds, but seems like the kill operation never went through
```
2024/08/06 14:49:22.831810  info clickhouse connection prepared: tcp://localhost:9000 run ping logger=clickhouse
2024/08/06 14:49:22.832816  warn clickhouse connection ping: tcp://localhost:9000 return error: code: 516, message: testt: Authentication failed: password is incorrect, or there is no user with such name., will wait 5 second to reconnect logger=clickhouse
2024/08/06 14:49:25.466613  info API call GET /backup/kill logger=server
2024/08/06 14:49:25.974107  info clickhouse connection prepared: tcp://localhost:9000 run ping logger=clickhouse
2024/08/06 14:49:25.976862  warn clickhouse connection ping: tcp://localhost:9000 return error: code: 516, message: testt: Authentication failed: password is incorrect, or there is no user with such name., will wait 5 second to reconnect logger=clickhouse
2024/08/06 14:49:27.841563  info clickhouse connection prepared: tcp://localhost:9000 run ping logger=clickhouse
2024/08/06 14:49:27.842560  warn clickhouse connection ping: tcp://localhost:9000 return error: code: 516, message: testt: Authentication failed: password is incorrect, or there is no user with such name., will wait 5 second to reconnect logger=clickhouse
2024/08/06 14:49:30.992981  info clickhouse connection prepared: tcp://localhost:9000 run ping logger=clickhouse
2024/08/06 14:49:30.994026  warn clickhouse connection ping: tcp://localhost:9000 return error: code: 516, message: testt: Authentication failed: password is incorrect, or there is no user with such name., will wait 5 second to reconnect logger=clickhouse
```
The db entry however, shows status=cancel and error='canceled from API /backup/kill'

**Reply by Slach** [reply]
Yep kill operation need to investigate, why context canceling doesn't work as expected
what is your `remote_storage` value?

**Reply by manasmulay** [reply]
Not using any remote storage in this operation, just a local backup `remote_storage: none`
